src/type-resolution/callables.ts

```diff
@@ -321,6 +326,61 @@ export function resolvePositionalCallableParameters(
   };
 }
 
+/**
+ * Returns an array of parameters such that scope and id (if it applies) are in
+ * the right positions.
+ */
+function prepareParameters(
+  x: ArrayLiteral,
+  callable: reflect.Callable,
+  logicalId?: string
```

Comment:
This is the place where the logical id is used. Every other function in this PR that got `logicalId` added to its signature only passes it down in order to reach this point. I'm not very happy with this situation, but I haven't found a good way to solve it.

One option I considered was to not receive the logical id here and, instead, inject a new pseudo-parameter (something like `CDK::ImplicitId`) in the second argument. Then, at the point where the lazy resource is resolved, this pseudo-parameter would be replaced with the id. But this would obscure the logic by creating a tight coupling between two functions that know nothing about each other.

Reply:
I'm not sure which way is better. But I do share your feelings about it being passed down all the way.

**Re: pseudo-parameter.** I don't think it's too bad to do this. Also it would not have to be a pseudo-parameter, but could be an AST type `lazy-logical-id`. We already have lazy evaluated resources, so why not have a lazy id?

Another option would be in **introduce a context to type resolution**. We will likely need something like it for #246 anyway. Basically a generic means of keeping track of where in the AST we are, what logicalID, what property, what the parent type is, etc.